cpp/src/io/parquet/decode_preprocess.cu

```diff
@@ -40,26 +41,136 @@ constexpr int rolling_buf_size = LEVEL_DECODE_BUF_SIZE;
+
+  size_t total_bytes = 0;
+  if (db != nullptr) {
+    // initialize with first value (which is stored in last_value)
+    if (lane_id == 0) { total_bytes = db->last_value; }
+
+    uleb128_t lane_sum = 0;
+    while (db->current_value_idx < db->num_encoded_values(true)) {
+      // calculate values for current mini-block
+      db->calc_mini_block_values(lane_id);
+
+      // get per lane sum for mini-block
+      for (uint32_t i = 0; i < db->values_per_mb; i += 32) {
+        uint32_t const idx = db->current_value_idx + i + lane_id;
+        if (idx < db->value_count) {
+          lane_sum += db->value[rolling_index<delta_rolling_buf_size>(idx)];
+        }
+      }
+
+      if (lane_id == 0) { db->setup_next_mini_block(true); }
+      __syncwarp();
+    }
+
+    // get sum for warp.
+    // note: warp_sum will only be valid on lane 0.
+    auto const warp_sum = WarpReduce(temp_storage[warp_id]).Sum(lane_sum);
+
+    if (lane_id == 0) { total_bytes += warp_sum; }
+  }
+  __syncthreads();
+
+  // now sum up total_bytes from the two warps. result is only valid on thread 0.
+  auto const final_bytes =
+    cudf::detail::single_lane_block_sum_reduce<preprocess_block_size, 0>(total_bytes);
+
+  return static_cast<size_type>(final_bytes);
+}
+
+/**
+ * @brief Calculate the number of string bytes in the page.
  *
  * This function expects the dictionary position to be at 0 and will traverse
- * the entire thing.
+ * the entire thing (for plain and dictionary encoding).
  *
- * Operates on a single warp only. Expects t < 32
+ * This expects all threads in the thread block (preprocess_block_size). Result is only
+ * valid on thread 0.
  *
  * @param s The local page info
  * @param t Thread index
  */
 __device__ size_type gpuDecodeTotalPageStringSize(page_state_s* s, int t)
 {
+  using cudf::detail::warp_size;
   size_type target_pos = s->num_input_values;
   size_type str_len    = 0;
-  if (s->dict_base) {
-    auto const [new_target_pos, len] =
-      gpuDecodeDictionaryIndices<true, unused_state_buf>(s, nullptr, target_pos, t);
-    target_pos = new_target_pos;
-    str_len    = len;
-  } else if ((s->col.data_type & 7) == BYTE_ARRAY) {
-    str_len = gpuInitStringDescriptors<true, unused_state_buf>(s, nullptr, target_pos, t);
+  switch (s->page.encoding) {
+    case Encoding::PLAIN_DICTIONARY:
+    case Encoding::RLE_DICTIONARY:
+      // TODO: make this block-based instead of just 1 warp
+      if (t < warp_size && s->dict_base) {
+        auto const [new_target_pos, len] =
+          gpuDecodeDictionaryIndices<true, unused_state_buf>(s, nullptr, target_pos, t);
+        target_pos = new_target_pos;
+        str_len    = len;
+      }
+      break;
+
+    case Encoding::PLAIN:
+      // TODO: since this is really just an estimate, we could just return
+      // s->dict_size (overestimate) or
+      // s->dict_size - sizeof(int) * s->page.num_input_values (underestimate)
```

Comment:
Am I mistaken that this was discussed and decided to find the real size?

Reply:
In my mind the discussion was more about the delta encodings...this is more a note that there could be a faster way to do this (esp since the plain string size calc is agonizingly slow). Unfortunately, at this point we don't really yet know how many values are present, so we can't get an exact number of string bytes. But if an overestimate is ok, we could just use `dict_size` and save a lot of time in this step.  Not sure if that's something to address with this PR.